### PR TITLE
types/lazy: add Peek method to SyncValue

### PR DESCRIPTION
**Pull request by bradfitz**
This adds the ability to "peek" at the value of a SyncValue, so that
it's possible to observe a value without computing this.

Updates tailscale/corp#17122

### CODE REVIEW

File: types/lazy/lazy.go [hunk 1]
```diff
@@ -41,15 +59,39 @@ func (z *SyncValue[T]) MustSet(val T) {
...
+	})
+	return z.v, *z.err.Load()
+}
+
+// Peek returns z's value and a boolean indicating whether the value has been
+// set. If a value has not been set, the zero value of T is returned.
+//
+// This function is safe to call concurrently with Get/GetErr/Set, but it's
+// undefined whether a value set by a concurrent call will be visible to Peek.
+func (z *SyncValue[T]) Peek() (T, bool) {
```
Comment (maisem) [inline types/lazy/lazy.go:89]: should this also return the error too?

Comment (maisem) [reply types/lazy/lazy.go:89]: maybe PeekErr as an additional func

**Review by maisem — APPROVED** [review]

File: types/lazy/lazy.go (hunk 1 above)
Comment (bradfitz) [reply types/lazy/lazy.go]: done

File: types/lazy/lazy.go [hunk 2]
```diff
@@ -4,7 +4,17 @@
...
+	"reflect"
+	"sync"
+	"sync/atomic"
+
+	"tailscale.com/types/ptr"
+)
+
+// nilErrPtr is a sentinel *error value for SyncValue.err to signal
+// that SyncValue.v is valid.
+var nilErrPtr = reflect.New(reflect.TypeFor[error]()).Interface().(*error)
```
Comment (andrew-d) [inline types/lazy/lazy.go:17]: OOC, why use `reflect.New` here?

File: types/lazy/lazy.go [hunk 3]
```diff
@@ -26,6 +43,7 @@ func (z *SyncValue[T]) Set(val T) bool {
 	var wasSet bool
 	z.once.Do(func() {
 		z.v = val
+		z.err.Store(nilErrPtr)
```
Comment (andrew-d) [inline types/lazy/lazy.go:46]: Comment that this must be after the assignment to `z.v`, otherwise this is racy?

**Review by andrew-d — APPROVED** [review]

File: types/lazy/lazy.go (hunk 2 above)
Comment (bradfitz) [reply types/lazy/lazy.go:17]: It's the first thing I thought of. What were you thinking of?

Comment (bradfitz) [reply types/lazy/lazy.go:17]: Oh, `var nilErrPtr = ptr.To((error)(nil))` works too

Comment (andrew-d) [reply types/lazy/lazy.go:17]: Yeah that 😃

Comment (bradfitz) [reply types/lazy/lazy.go:17]: or `var nilErrPtr = ptr.To[error](nil)` is a bit prettier

File: types/lazy/lazy.go (hunk 3 above)
Comment (bradfitz) [reply types/lazy/lazy.go:46]: done, in several spots

**Pull request merged**